--- /home/mars-user/office-open-repro/upstream/wine-8.0.1/dlls/ntdll/unix/file.c	2023-04-20 11:34:03.000000000 +0000
+++ /home/mars-user/office-open-repro/source-extract/sources/wine/dlls/ntdll/unix/file.c	2023-09-25 18:01:44.000000000 +0000
@@ -1292,18 +1292,24 @@
 /***********************************************************************
  *           is_hidden_file
  *
- * Check if the specified file should be hidden based on its name and the show dot files option.
+ * Check if the specified file should be hidden based on its unix path and the show dot files option.
  */
-static BOOL is_hidden_file( const UNICODE_STRING *name )
+static BOOL is_hidden_file( const char *name )
 {
-    WCHAR *p, *end;
+    const char *p;
 
     if (show_dot_files) return FALSE;
 
-    end = p = name->Buffer + name->Length/sizeof(WCHAR);
-    while (p > name->Buffer && p[-1] == '\\') p--;
-    while (p > name->Buffer && p[-1] != '\\') p--;
-    return (p < end && *p == '.');
+    p = name + strlen( name );
+    while (p > name && p[-1] == '/') p--;
+    while (p > name && p[-1] != '/') p--;
+    /* CrossOver Hack for bug 15207 - hide files starting in ~$ */
+    if (p[0] == '~' && p[1] == '$') return TRUE;
+    if (*p++ != '.') return FALSE;
+    if (!*p || *p == '/') return FALSE;  /* "." directory */
+    if (*p++ != '.') return TRUE;
+    if (!*p || *p == '/') return FALSE;  /* ".." directory */
+    return TRUE;
 }
 
 
@@ -1319,30 +1325,50 @@
 {
     static const char hash_chars[32] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ012345";
 
-    LPCWSTR p, ext, end = name + length;
+    LPCWSTR p, ext, hash_end, end = name + length;
     LPWSTR dst;
     unsigned short hash;
     int i;
 
+    /* Find last dot for start of the extension */
+    for (p = name + 1, ext = NULL; p < end - 1; p++) if (*p == '.') ext = p;
+
+    /* don't include the standard 3 char .ext in the filename hash */
+    hash_end = end;
+    if (ext && ((end - ext) == 4 ))
+    {
+        /*
+         * FIXME: CodeWeavers hack alert
+         * The next five lines are a nasty hack to only activate this
+         * (more correct behaviour) for Quicken files for the moment.
+         * We don't want to break our install base of programs that have
+         * shortfile names stored in the registry or elsewhere.
+         */
+        WCHAR szqdf[]={'.','q','d','f'};
+        WCHAR szqsd[]={'.','q','s','d'};
+        WCHAR szqel[]={'.','q','e','l'};
+        WCHAR szqph[]={'.','q','p','h'};
+        if (!wcsnicmp(ext,szqdf,4) || !wcsnicmp(ext,szqsd,4) ||
+            !wcsnicmp(ext,szqph,4) || !wcsnicmp(ext,szqel,4))
+            hash_end = ext;
+    }
+
     /* Compute the hash code of the file name */
     /* If you know something about hash functions, feel free to */
     /* insert a better algorithm here... */
     if (!is_case_sensitive)
     {
-        for (p = name, hash = 0xbeef; p < end - 1; p++)
+        for (p = name, hash = 0xbeef; p < hash_end - 1; p++)
             hash = (hash<<3) ^ (hash>>5) ^ towlower(*p) ^ (towlower(p[1]) << 8);
         hash = (hash<<3) ^ (hash>>5) ^ towlower(*p); /* Last character */
     }
     else
     {
-        for (p = name, hash = 0xbeef; p < end - 1; p++)
+        for (p = name, hash = 0xbeef; p < hash_end - 1; p++)
             hash = (hash << 3) ^ (hash >> 5) ^ *p ^ (p[1] << 8);
         hash = (hash << 3) ^ (hash >> 5) ^ *p;  /* Last character */
     }
 
-    /* Find last dot for start of the extension */
-    for (p = name + 1, ext = NULL; p < end - 1; p++) if (*p == '.') ext = p;
-
     /* Copy first 4 chars, replacing invalid chars with '_' */
     for (i = 4, p = name, dst = buffer; i > 0; i--, p++)
     {
@@ -1600,11 +1626,11 @@
 }
 
 
-static int fd_set_dos_attrib( int fd, UINT attr )
+static int fd_set_dos_attrib( int fd, UINT attr, BOOL force_set )
 {
     /* we only store the HIDDEN and SYSTEM attributes */
     attr &= XATTR_ATTRIBS_MASK;
-    if (attr != 0)
+    if (force_set || attr != 0)
     {
         /* encode the attributes in Samba 3 ASCII format. Samba 4 has extended
          * this format with more features, but retains compatibility with the
@@ -1618,7 +1644,7 @@
 
 
 /* set the stat info and file attributes for a file (by file descriptor) */
-NTSTATUS fd_set_file_info( int fd, UINT attr )
+static NTSTATUS fd_set_file_info( int fd, UINT attr, BOOL force_set_xattr )
 {
     struct stat st;
 
@@ -1637,7 +1663,11 @@
     }
     if (fchmod( fd, st.st_mode ) == -1) return errno_to_status( errno );
 
-    if (fd_set_dos_attrib( fd, attr ) == -1 && errno != ENOTSUP)
+    /* if the file has multiple names, we can't be sure that it is safe to not
+       set the extended attribute, since any of the names could start with a dot */
+    force_set_xattr = force_set_xattr || st.st_nlink > 1;
+
+    if (fd_set_dos_attrib( fd, attr, force_set_xattr ) == -1 && errno != ENOTSUP)
         WARN( "Failed to set extended attribute " SAMBA_XATTR_DOS_ATTRIB ". errno %d (%s)\n",
               errno, strerror( errno ) );
 
@@ -1682,6 +1712,8 @@
         *attr |= parse_samba_dos_attrib_data( attr_data, attr_len );
     else
     {
+        if (is_hidden_file( path ))
+            *attr |= FILE_ATTRIBUTE_HIDDEN;
         if (errno == ENOTSUP) return ret;
 #ifdef ENODATA
         if (errno == ENODATA) return ret;
@@ -2176,7 +2208,6 @@
     OBJECT_ATTRIBUTES attr;
     UNICODE_STRING string;
     char *unix_name;
-    IO_STATUS_BLOCK io = {{0}};
     HANDLE mountmgr;
     unsigned int status;
     int letter;
@@ -2202,8 +2233,7 @@
                                       FILE_SHARE_READ | FILE_SHARE_WRITE, FILE_SYNCHRONOUS_IO_NONALERT );
     if (status) return status;
 
-    status = NtDeviceIoControlFile( mountmgr, NULL, NULL, NULL, &io, IOCTL_MOUNTMGR_QUERY_UNIX_DRIVE,
-                                    drive, sizeof(*drive), drive, size );
+    status = sync_ioctl( mountmgr, IOCTL_MOUNTMGR_QUERY_UNIX_DRIVE, drive, sizeof(*drive), drive, size );
     NtClose( mountmgr );
     if (status == STATUS_BUFFER_OVERFLOW) status = STATUS_SUCCESS;
     else if (status) WARN("failed to retrieve filesystem type from mountmgr, status %#x\n", status);
@@ -2216,7 +2246,7 @@
  *
  * Return a directory entry from the cached data.
  */
-static NTSTATUS get_dir_data_entry( struct dir_data *dir_data, void *info_ptr, IO_STATUS_BLOCK *io,
+static NTSTATUS get_dir_data_entry( struct dir_data *dir_data, void *info_ptr, unsigned int *used_size,
                                     ULONG max_length, FILE_INFORMATION_CLASS class,
                                     union file_directory_info **last_info )
 {
@@ -2235,7 +2265,7 @@
         TRACE( "ignoring file %s\n", names->unix_name );
         return STATUS_SUCCESS;
     }
-    start = dir_info_align( io->Information );
+    start = dir_info_align( *used_size );
     dir_size = dir_info_size( class, 0 );
     if (start + dir_size > max_length) return STATUS_MORE_ENTRIES;
 
@@ -2252,11 +2282,6 @@
     if (class != FileNamesInformation)
     {
         if (st.st_dev != dir_data->id.dev) st.st_ino = 0;  /* ignore inode if on a different device */
-
-        if (!show_dot_files && names->long_name[0] == '.' && names->long_name[1] &&
-            (names->long_name[1] != '.' || names->long_name[2]))
-            attributes |= FILE_ATTRIBUTE_HIDDEN;
-
         fill_file_info( &st, attributes, info, class );
     }
 
@@ -2305,7 +2330,7 @@
     }
 
     memcpy( (char *)info + dir_size, names->long_name, min( name_len, max_length ) );
-    io->Information = start + dir_size + min( name_len, max_length );
+    *used_size = start + dir_size + min( name_len, max_length );
     if (*last_info) (*last_info)->next = (char *)info - (char *)*last_info;
     *last_info = info;
     return name_len > max_length ? STATUS_BUFFER_OVERFLOW : STATUS_SUCCESS;
@@ -2604,17 +2629,19 @@
  *              NtQueryDirectoryFile   (NTDLL.@)
  */
 NTSTATUS WINAPI NtQueryDirectoryFile( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc_routine,
-                                      void *apc_context, IO_STATUS_BLOCK *io, void *buffer, ULONG length,
+                                      void *apc_context, IO_STATUS_BLOCK *iosb, void *buffer, ULONG length,
                                       FILE_INFORMATION_CLASS info_class, BOOLEAN single_entry,
                                       UNICODE_STRING *mask, BOOLEAN restart_scan )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
+    unsigned int used_size = 0;
     int cwd, fd, needs_close;
     enum server_fd_type type;
     struct dir_data *data;
     unsigned int status;
 
     TRACE("(%p %p %p %p %p %p 0x%08x 0x%08x 0x%08x %s 0x%08x\n",
-          handle, event, apc_routine, apc_context, io, buffer,
+          handle, event, apc_routine, apc_context, iosb, buffer,
           (int)length, info_class, single_entry, debugstr_us(mask),
           restart_scan);
 
@@ -2657,8 +2684,6 @@
         return STATUS_INVALID_PARAMETER;
     }
 
-    io->Information = 0;
-
     mutex_lock( &dir_mutex );
 
     cwd = open( ".", O_RDONLY );
@@ -2672,7 +2697,7 @@
 
             while (!status && data->pos < data->count)
             {
-                status = get_dir_data_entry( data, buffer, io, length, info_class, &last_info );
+                status = get_dir_data_entry( data, buffer, &used_size, length, info_class, &last_info );
                 if (!status || status == STATUS_BUFFER_OVERFLOW) data->pos++;
                 if (single_entry && last_info) break;
             }
@@ -2680,7 +2705,7 @@
             if (!last_info) status = STATUS_NO_MORE_FILES;
             else if (status == STATUS_MORE_ENTRIES) status = STATUS_SUCCESS;
 
-            io->u.Status = status;
+            set_async_iosb( io, status, used_size );
         }
         if (cwd == -1 || fchdir( cwd ) == -1) chdir( "/" );
     }
@@ -2690,7 +2715,7 @@
 
     if (needs_close) close( fd );
     if (cwd != -1) close( cwd );
-    TRACE( "=> %x (%ld)\n", status, io->Information );
+    TRACE( "=> %#x (%u)\n", status, used_size );
     return status;
 }
 
@@ -3966,6 +3991,7 @@
     OBJECT_ATTRIBUTES new_attr;
     UNICODE_STRING nt_name;
     char *unix_name;
+    BOOL name_hidden = FALSE;
     BOOL created = FALSE;
     unsigned int status;
 
@@ -4008,6 +4034,7 @@
 
     if (status == STATUS_SUCCESS)
     {
+        name_hidden = is_hidden_file( unix_name );
         status = open_unix_file( handle, unix_name, access, &new_attr, attributes,
                                  sharing, disposition, options, ea_buffer, ea_length );
         free( unix_name );
@@ -4035,14 +4062,15 @@
             break;
         }
 
-        if (io->Information == FILE_CREATED && (attributes & XATTR_ATTRIBS_MASK))
+        if (io->Information == FILE_CREATED &&
+            ((attributes & XATTR_ATTRIBS_MASK) || name_hidden))
         {
             int fd, needs_close;
 
             /* set any DOS extended attributes */
             if (!server_get_unix_fd( *handle, 0, &fd, &needs_close, NULL, NULL ))
             {
-                if (fd_set_dos_attrib( fd, attributes ) == -1 && errno != ENOTSUP)
+                if (fd_set_dos_attrib( fd, attributes, TRUE ) == -1 && errno != ENOTSUP)
                     WARN( "Failed to set extended attribute " SAMBA_XATTR_DOS_ATTRIB ". errno %d (%s)",
                           errno, strerror( errno ) );
                 if (needs_close) close( fd );
@@ -4213,7 +4241,6 @@
             info->AllocationSize = std.AllocationSize;
             info->EndOfFile      = std.EndOfFile;
             info->FileAttributes = basic.FileAttributes;
-            if (is_hidden_file( attr->ObjectName )) info->FileAttributes |= FILE_ATTRIBUTE_HIDDEN;
         }
         free( unix_name );
     }
@@ -4244,10 +4271,7 @@
         else if (!S_ISREG(st.st_mode) && !S_ISDIR(st.st_mode))
             status = STATUS_INVALID_INFO_CLASS;
         else
-        {
             status = fill_file_info( &st, attributes, info, FileBasicInformation );
-            if (is_hidden_file( attr->ObjectName )) info->FileAttributes |= FILE_ATTRIBUTE_HIDDEN;
-        }
         free( unix_name );
     }
     else WARN( "%s not found (%x)\n", debugstr_us(attr->ObjectName), status );
@@ -4563,20 +4587,25 @@
         {
             const FILE_BASIC_INFORMATION *info = ptr;
             LARGE_INTEGER mtime, atime;
+            char *unix_name;
 
             if ((status = server_get_unix_fd( handle, 0, &fd, &needs_close, NULL, NULL )))
                 return io->u.Status = status;
 
+            if (server_get_unix_name( handle, &unix_name )) unix_name = NULL;
+
             mtime.QuadPart = info->LastWriteTime.QuadPart == -1 ? 0 : info->LastWriteTime.QuadPart;
             atime.QuadPart = info->LastAccessTime.QuadPart == -1 ? 0 : info->LastAccessTime.QuadPart;
 
             if (atime.QuadPart || mtime.QuadPart)
                 status = set_file_times( fd, &mtime, &atime );
 
-            if (status == STATUS_SUCCESS && info->FileAttributes)
-                status = fd_set_file_info( fd, info->FileAttributes );
+            if (status == STATUS_SUCCESS)
+                status = fd_set_file_info( fd, info->FileAttributes,
+                                           unix_name && is_hidden_file( unix_name ));
 
             if (needs_close) close( fd );
+            free( unix_name );
         }
         else status = STATUS_INVALID_PARAMETER_3;
         break;
@@ -5018,7 +5047,7 @@
 
 /* do a read call through the server */
 static unsigned int server_read_file( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_context,
-                                      IO_STATUS_BLOCK *io, void *buffer, ULONG size,
+                                      client_ptr_t io, void *buffer, ULONG size,
                                       LARGE_INTEGER *offset, ULONG *key )
 {
     struct async_irp *async;
@@ -5034,17 +5063,14 @@
 
     SERVER_START_REQ( read )
     {
-        req->async = server_async( handle, &async->io, event, apc, apc_context, iosb_client_ptr(io) );
+        req->async = server_async( handle, &async->io, event, apc, apc_context, io );
         req->pos   = offset ? offset->QuadPart : 0;
         wine_server_set_reply( req, buffer, size );
         status = virtual_locked_server_call( req );
         wait_handle = wine_server_ptr_handle( reply->wait );
         options     = reply->options;
         if (wait_handle && status != STATUS_PENDING)
-        {
-            io->u.Status    = status;
-            io->Information = wine_server_reply_size( reply );
-        }
+            set_async_iosb( io, status, wine_server_reply_size( reply ) );
     }
     SERVER_END_REQ;
 
@@ -5056,7 +5082,7 @@
 
 /* do a write call through the server */
 static unsigned int server_write_file( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_context,
-                                       IO_STATUS_BLOCK *io, const void *buffer, ULONG size,
+                                       client_ptr_t io, const void *buffer, ULONG size,
                                        LARGE_INTEGER *offset, ULONG *key )
 {
     struct async_irp *async;
@@ -5072,17 +5098,14 @@
 
     SERVER_START_REQ( write )
     {
-        req->async = server_async( handle, &async->io, event, apc, apc_context, iosb_client_ptr(io) );
+        req->async = server_async( handle, &async->io, event, apc, apc_context, io );
         req->pos   = offset ? offset->QuadPart : 0;
         wine_server_add_data( req, buffer, size );
         status = wine_server_call( req );
         wait_handle = wine_server_ptr_handle( reply->wait );
         options     = reply->options;
         if (wait_handle && status != STATUS_PENDING)
-        {
-            io->u.Status    = status;
-            io->Information = reply->size;
-        }
+            set_async_iosb( io, status, reply->size );
     }
     SERVER_END_REQ;
 
@@ -5095,7 +5118,7 @@
 /* do an ioctl call through the server */
 static NTSTATUS server_ioctl_file( HANDLE handle, HANDLE event,
                                    PIO_APC_ROUTINE apc, PVOID apc_context,
-                                   IO_STATUS_BLOCK *io, UINT code,
+                                   client_ptr_t io, UINT code,
                                    const void *in_buffer, UINT in_size,
                                    PVOID out_buffer, UINT out_size )
 {
@@ -5112,7 +5135,7 @@
     SERVER_START_REQ( ioctl )
     {
         req->code        = code;
-        req->async       = server_async( handle, &async->io, event, apc, apc_context, iosb_client_ptr(io) );
+        req->async       = server_async( handle, &async->io, event, apc, apc_context, io );
         wine_server_add_data( req, in_buffer, in_size );
         if ((code & 3) != METHOD_BUFFERED) wine_server_add_data( req, out_buffer, out_size );
         wine_server_set_reply( req, out_buffer, out_size );
@@ -5120,10 +5143,7 @@
         wait_handle = wine_server_ptr_handle( reply->wait );
         options     = reply->options;
         if (wait_handle && status != STATUS_PENDING)
-        {
-            io->u.Status    = status;
-            io->Information = wine_server_reply_size( reply );
-        }
+            set_async_iosb( io, status, wine_server_reply_size( reply ));
     }
     SERVER_END_REQ;
 
@@ -5149,8 +5169,6 @@
 static unsigned int get_io_timeouts( HANDLE handle, enum server_fd_type type, ULONG count, BOOL is_read,
                                      struct io_timeouts *timeouts )
 {
-    unsigned int status = STATUS_SUCCESS;
-
     timeouts->interval = timeouts->total = -1;
 
     switch(type)
@@ -5159,11 +5177,9 @@
     {
         /* GetCommTimeouts */
         SERIAL_TIMEOUTS st;
-        IO_STATUS_BLOCK io = {{0}};
 
-        status = NtDeviceIoControlFile( handle, NULL, NULL, NULL, &io,
-                                        IOCTL_SERIAL_GET_TIMEOUTS, NULL, 0, &st, sizeof(st) );
-        if (status) break;
+        if (sync_ioctl( handle, IOCTL_SERIAL_GET_TIMEOUTS, NULL, 0, &st, sizeof(st) ))
+            break;
 
         if (is_read)
         {
@@ -5198,8 +5214,7 @@
             {
                 req->handle = wine_server_obj_handle( handle );
                 req->flags = 0;
-                if (!(status = wine_server_call( req )) &&
-                    reply->read_timeout != TIMEOUT_INFINITE)
+                if (!wine_server_call( req ) && reply->read_timeout != TIMEOUT_INFINITE)
                     timeouts->total = reply->read_timeout / -10000;
             }
             SERVER_END_REQ;
@@ -5246,14 +5261,13 @@
     {
         /* GetCommTimeouts */
         SERIAL_TIMEOUTS st;
-        IO_STATUS_BLOCK io = {{0}};
 
-        status = NtDeviceIoControlFile( handle, NULL, NULL, NULL, &io,
-                                        IOCTL_SERIAL_GET_TIMEOUTS, NULL, 0, &st, sizeof(st) );
-        if (status) break;
-        *avail_mode = (!st.ReadTotalTimeoutMultiplier &&
-                       !st.ReadTotalTimeoutConstant &&
-                       st.ReadIntervalTimeout == MAXDWORD);
+        if (!(status = sync_ioctl( handle, IOCTL_SERIAL_GET_TIMEOUTS, NULL, 0, &st, sizeof(st) )))
+        {
+            *avail_mode = (!st.ReadTotalTimeoutMultiplier &&
+                           !st.ReadTotalTimeoutConstant &&
+                           st.ReadIntervalTimeout == MAXDWORD);
+        }
         break;
     }
     case FD_TYPE_MAILSLOT:
@@ -5326,26 +5340,65 @@
     return status;
 }
 
+static BOOL is_quickenpatch(void)
+{
+    static const WCHAR qkn[] = {'q','u','i','c','k','e','n','P','a','t','c','h','.','e','x','e',0};
+    WCHAR *path = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
+    DWORD len = sizeof(qkn)/sizeof(qkn[0]) - 1, len2 = wcslen(path);
+    return (len <= len2 && !wcsicmp( path + len2 - len, qkn ));
+}
+
+static BOOL is_red_launcher(void)
+{
+    static const WCHAR red_launcher[] = {'R','E','D','L','a','u','n','c','h','e','r','.','e','x','e',0};
+    static const WCHAR red_prelauncher[] = {'R','E','D','p','r','e','l','a','u','n','c','h','e','r','.','e','x','e',0};
+    WCHAR *path = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
+    DWORD path_len = wcslen(path);
+
+    DWORD len = sizeof(red_launcher) / sizeof(red_launcher[0]) - 1;
+    if (len <= path_len && !wcsicmp( path + path_len - len, red_launcher ))
+        return TRUE;
+
+    len = sizeof(red_prelauncher) / sizeof(red_prelauncher[0]) - 1;
+    return len <= path_len && !wcsicmp( path + path_len - len, red_prelauncher );
+}
+
+static void replace_dx12_string(char *buffer, ULONG length)
+{
+    static const char dx12_string[] = "\"fallback\": \"DirectX 12\"";
+    static const char dx11_string[] = "\"fallback\": \"DirectX 11\"";
+    char *dst = strstr(buffer, dx12_string);
+    if (dst)
+        memcpy(dst, dx11_string, sizeof(dx11_string) - 1);
+}
+
+/* CW HACK 14391 */
+NTSTATUS WINAPI __wine_rpc_NtReadFile( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_user,
+                                IO_STATUS_BLOCK *io, void *buffer, ULONG length,
+                                LARGE_INTEGER *offset, ULONG *key )
+{
+    return NtReadFile( handle, event, apc, apc_user, io, buffer, length, offset, key );
+}
 
 /******************************************************************************
  *              NtReadFile   (NTDLL.@)
  */
 NTSTATUS WINAPI NtReadFile( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_user,
-                            IO_STATUS_BLOCK *io, void *buffer, ULONG length,
+                            IO_STATUS_BLOCK *iosb, void *buffer, ULONG length,
                             LARGE_INTEGER *offset, ULONG *key )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
     int result, unix_handle, needs_close;
     unsigned int options;
     struct io_timeouts timeouts;
     unsigned int status, ret_status;
     UINT total = 0;
-    client_ptr_t iosb_ptr = iosb_client_ptr(io);
     enum server_fd_type type;
     ULONG_PTR cvalue = apc ? 0 : (ULONG_PTR)apc_user;
     BOOL send_completion = FALSE, async_read, timeout_init_done = FALSE;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p)\n",
-           handle, event, apc, apc_user, io, buffer, (int)length, offset, key );
+           handle, event, apc, apc_user, iosb, buffer, (int)length, offset, key );
 
     if (!io) return STATUS_ACCESS_VIOLATION;
 
@@ -5372,6 +5425,9 @@
             /* async I/O doesn't make sense on regular files */
             while ((result = virtual_locked_pread( unix_handle, buffer, length, offset->QuadPart )) == -1)
             {
+                /* CrossOver hack 14664 */
+                if (errno == EFAULT && is_quickenpatch() && virtual_check_buffer_for_write( buffer, length ))
+                    continue;
                 if (errno != EINTR)
                 {
                     status = errno_to_status( errno );
@@ -5409,7 +5465,7 @@
         if ((status = get_io_timeouts( handle, type, length, TRUE, &timeouts ))) goto err;
         if (timeouts.interval)
         {
-            status = register_async_file_read( handle, event, apc, apc_user, iosb_ptr,
+            status = register_async_file_read( handle, event, apc, apc_user, io,
                                                buffer, total, length, FALSE );
             goto err;
         }
@@ -5451,6 +5507,9 @@
         else if (errno != EAGAIN)
         {
             if (errno == EINTR) continue;
+            /* CrossOver hack 14664 */
+            if (errno == EFAULT && is_quickenpatch() && virtual_check_buffer_for_write( buffer, length ))
+                continue;
             if (!total) status = errno_to_status( errno );
             goto err;
         }
@@ -5465,7 +5524,7 @@
                 status = STATUS_SUCCESS;
                 goto done;
             }
-            status = register_async_file_read( handle, event, apc, apc_user, iosb_ptr,
+            status = register_async_file_read( handle, event, apc, apc_user, io,
                                                buffer, total, length, avail_mode );
             goto err;
         }
@@ -5509,12 +5568,14 @@
     if (needs_close) close( unix_handle );
     if (status == STATUS_SUCCESS || (status == STATUS_END_OF_FILE && (!async_read || type == FD_TYPE_FILE)))
     {
-        io->u.Status = status;
-        io->Information = total;
+        set_async_iosb( io, status, total );
         TRACE("= SUCCESS (%u)\n", total);
+        /* CrossOver hack for bug 21786 */
+        if (!apc && !event && length == 0x1000 && status == STATUS_SUCCESS && is_red_launcher())
+            replace_dx12_string(buffer, length);
         if (event) NtSetEvent( event, NULL );
         if (apc && (!status || async_read)) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc,
-                                                              (ULONG_PTR)apc_user, iosb_ptr, 0 );
+                                                              (ULONG_PTR)apc_user, io, 0 );
     }
     else
     {
@@ -5534,19 +5595,19 @@
  *              NtReadFileScatter   (NTDLL.@)
  */
 NTSTATUS WINAPI NtReadFileScatter( HANDLE file, HANDLE event, PIO_APC_ROUTINE apc, void *apc_user,
-                                   IO_STATUS_BLOCK *io, FILE_SEGMENT_ELEMENT *segments,
+                                   IO_STATUS_BLOCK *iosb, FILE_SEGMENT_ELEMENT *segments,
                                    ULONG length, LARGE_INTEGER *offset, ULONG *key )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
     int result, unix_handle, needs_close;
     unsigned int options, status;
     UINT pos = 0, total = 0;
-    client_ptr_t iosb_ptr = iosb_client_ptr(io);
     enum server_fd_type type;
     ULONG_PTR cvalue = apc ? 0 : (ULONG_PTR)apc_user;
     BOOL send_completion = FALSE;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p),partial stub!\n",
-           file, event, apc, apc_user, io, segments, (int)length, offset, key );
+           file, event, apc, apc_user, iosb, segments, (int)length, offset, key );
 
     if (!io) return STATUS_ACCESS_VIOLATION;
 
@@ -5590,11 +5651,10 @@
     send_completion = cvalue != 0;
 
     if (needs_close) close( unix_handle );
-    io->u.Status = status;
-    io->Information = total;
+    set_async_iosb( io, status, total );
     TRACE("= 0x%08x (%u)\n", status, total);
     if (event) NtSetEvent( event, NULL );
-    if (apc) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc, (ULONG_PTR)apc_user, iosb_ptr, 0 );
+    if (apc) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc, (ULONG_PTR)apc_user, io, 0 );
     if (send_completion) add_completion( file, cvalue, status, total, TRUE );
 
     return STATUS_PENDING;
@@ -5611,22 +5671,22 @@
  *              NtWriteFile   (NTDLL.@)
  */
 NTSTATUS WINAPI NtWriteFile( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_user,
-                             IO_STATUS_BLOCK *io, const void *buffer, ULONG length,
+                             IO_STATUS_BLOCK *iosb, const void *buffer, ULONG length,
                              LARGE_INTEGER *offset, ULONG *key )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
     int result, unix_handle, needs_close;
     unsigned int options;
     struct io_timeouts timeouts;
     unsigned int status, ret_status;
     UINT total = 0;
-    client_ptr_t iosb_ptr = iosb_client_ptr(io);
     enum server_fd_type type;
     ULONG_PTR cvalue = apc ? 0 : (ULONG_PTR)apc_user;
     BOOL send_completion = FALSE, async_write, append_write = FALSE, timeout_init_done = FALSE;
     LARGE_INTEGER offset_eof;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p)\n",
-           handle, event, apc, apc_user, io, buffer, (int)length, offset, key );
+           handle, event, apc, apc_user, iosb, buffer, (int)length, offset, key );
 
     if (!io) return STATUS_ACCESS_VIOLATION;
 
@@ -5768,7 +5828,7 @@
             {
                 req->type   = ASYNC_TYPE_WRITE;
                 req->count  = length;
-                req->async  = server_async( handle, &fileio->io, event, apc, apc_user, iosb_ptr );
+                req->async  = server_async( handle, &fileio->io, event, apc, apc_user, io );
                 status = wine_server_call( req );
             }
             SERVER_END_REQ;
@@ -5819,11 +5879,10 @@
 
     if (status == STATUS_SUCCESS)
     {
-        io->u.Status = status;
-        io->Information = total;
+        set_async_iosb( io, status, total );
         TRACE("= SUCCESS (%u)\n", total);
         if (event) NtSetEvent( event, NULL );
-        if (apc) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc, (ULONG_PTR)apc_user, iosb_ptr, 0 );
+        if (apc) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc, (ULONG_PTR)apc_user, io, 0 );
     }
     else
     {
@@ -5841,19 +5900,19 @@
  *              NtWriteFileGather   (NTDLL.@)
  */
 NTSTATUS WINAPI NtWriteFileGather( HANDLE file, HANDLE event, PIO_APC_ROUTINE apc, void *apc_user,
-                                   IO_STATUS_BLOCK *io, FILE_SEGMENT_ELEMENT *segments,
+                                   IO_STATUS_BLOCK *iosb, FILE_SEGMENT_ELEMENT *segments,
                                    ULONG length, LARGE_INTEGER *offset, ULONG *key )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
     int result, unix_handle, needs_close;
     unsigned int options, status;
     UINT pos = 0, total = 0;
-    client_ptr_t iosb_ptr = iosb_client_ptr(io);
     enum server_fd_type type;
     ULONG_PTR cvalue = apc ? 0 : (ULONG_PTR)apc_user;
     BOOL send_completion = FALSE;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p),partial stub!\n",
-           file, event, apc, apc_user, io, segments, (int)length, offset, key );
+           file, event, apc, apc_user, iosb, segments, (int)length, offset, key );
 
     if (length % page_size) return STATUS_INVALID_PARAMETER;
     if (!io) return STATUS_ACCESS_VIOLATION;
@@ -5908,11 +5967,10 @@
     if (needs_close) close( unix_handle );
     if (status == STATUS_SUCCESS)
     {
-        io->u.Status = status;
-        io->Information = total;
+        set_async_iosb( io, status, total );
         TRACE("= SUCCESS (%u)\n", total);
         if (event) NtSetEvent( event, NULL );
-        if (apc) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc, (ULONG_PTR)apc_user, iosb_ptr, 0 );
+        if (apc) NtQueueApcThread( GetCurrentThread(), (PNTAPCFUNC)apc, (ULONG_PTR)apc_user, io, 0 );
     }
     else
     {
@@ -5928,14 +5986,15 @@
  *              NtDeviceIoControlFile   (NTDLL.@)
  */
 NTSTATUS WINAPI NtDeviceIoControlFile( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_context,
-                                       IO_STATUS_BLOCK *io, ULONG code, void *in_buffer, ULONG in_size,
+                                       IO_STATUS_BLOCK *iosb, ULONG code, void *in_buffer, ULONG in_size,
                                        void *out_buffer, ULONG out_size )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
     ULONG device = (code >> 16);
     NTSTATUS status = STATUS_NOT_SUPPORTED;
 
     TRACE( "(%p,%p,%p,%p,%p,0x%08x,%p,0x%08x,%p,0x%08x)\n",
-           handle, event, apc, apc_context, io, (int)code,
+           handle, event, apc, apc_context, iosb, (int)code,
            in_buffer, (int)in_size, out_buffer, (int)out_size );
 
     /* some broken applications call this frequently with INVALID_HANDLE_VALUE,
@@ -5972,11 +6031,24 @@
         return server_ioctl_file( handle, event, apc, apc_context, io, code,
                                   in_buffer, in_size, out_buffer, out_size );
 
-    if (status != STATUS_PENDING && !NT_ERROR(status)) io->u.Status = status;
+    if (status != STATUS_PENDING && !NT_ERROR(status))
+        set_async_iosb( io, status, 0 );
     return status;
 }
 
 
+/* helper for internal ioctl calls */
+NTSTATUS sync_ioctl( HANDLE file, ULONG code, void *in_buffer, ULONG in_size, void *out_buffer, ULONG out_size )
+{
+    IO_STATUS_BLOCK32 io32;
+    IO_STATUS_BLOCK io;
+
+    /* the 32-bit iosb is filled for overlapped file handles */
+    io.u.Pointer = &io32;
+    return NtDeviceIoControlFile( file, NULL, NULL, NULL, &io, code, in_buffer, in_size, out_buffer, out_size );
+}
+
+
 /* Tell Valgrind to ignore any holes in structs we will be passing to the
  * server */
 static void ignore_server_ioctl_struct_holes( ULONG code, const void *in_buffer, ULONG in_size )
@@ -6005,13 +6077,15 @@
  *              NtFsControlFile   (NTDLL.@)
  */
 NTSTATUS WINAPI NtFsControlFile( HANDLE handle, HANDLE event, PIO_APC_ROUTINE apc, void *apc_context,
-                                 IO_STATUS_BLOCK *io, ULONG code, void *in_buffer, ULONG in_size,
+                                 IO_STATUS_BLOCK *iosb, ULONG code, void *in_buffer, ULONG in_size,
                                  void *out_buffer, ULONG out_size )
 {
+    client_ptr_t io = iosb_client_ptr(iosb);
+    ULONG_PTR information = 0;
     NTSTATUS status;
 
     TRACE( "(%p,%p,%p,%p,%p,0x%08x,%p,0x%08x,%p,0x%08x)\n",
-           handle, event, apc, apc_context, io, (int)code,
+           handle, event, apc, apc_context, iosb, (int)code,
            in_buffer, (int)in_size, out_buffer, (int)out_size );
 
     if (!io) return STATUS_INVALID_PARAMETER;
@@ -6051,19 +6125,19 @@
             buffer->StartingVcn.QuadPart        = 1;
             buffer->Extents[0].NextVcn.QuadPart = 0;
             buffer->Extents[0].Lcn.QuadPart     = 0;
-            io->Information = sizeof(RETRIEVAL_POINTERS_BUFFER);
+            information = sizeof(RETRIEVAL_POINTERS_BUFFER);
             status = STATUS_SUCCESS;
         }
         else
         {
-            io->Information = 0;
+            information = 0;
             status = STATUS_BUFFER_TOO_SMALL;
         }
         break;
     }
 
     case FSCTL_GET_REPARSE_POINT:
-        io->Information = 0;
+        information = 0;
         if (out_buffer && out_size)
         {
             FIXME("FSCTL_GET_REPARSE_POINT semi-stub\n");
@@ -6078,7 +6152,7 @@
         int fd, needs_close;
         struct stat st;
 
-        io->Information = 0;
+        information = 0;
         if (out_size >= sizeof(*info))
         {
             status = server_get_unix_fd( handle, 0, &fd, &needs_close, NULL, NULL );
@@ -6088,7 +6162,7 @@
             memset( info, 0, sizeof(*info) );
             memcpy( info->ObjectId, &st.st_dev, sizeof(st.st_dev) );
             memcpy( info->ObjectId + 8, &st.st_ino, sizeof(st.st_ino) );
-            io->Information = sizeof(*info);
+            information = sizeof(*info);
         }
         else status = STATUS_BUFFER_TOO_SMALL;
         break;
@@ -6096,7 +6170,7 @@
 
     case FSCTL_SET_SPARSE:
         TRACE("FSCTL_SET_SPARSE: Ignoring request\n");
-        io->Information = 0;
+        information = 0;
         status = STATUS_SUCCESS;
         break;
     default:
@@ -6104,7 +6178,8 @@
                                   in_buffer, in_size, out_buffer, out_size );
     }
 
-    if (status != STATUS_PENDING) io->u.Status = status;
+    if (status != STATUS_PENDING)
+        set_async_iosb( io, status, information );
     return status;
 }
 
